File: Ejercicios/Ejercicio3/3.py
from abc import ABC, abstractmethod

class Empleado(ABC):
    def __init__(self, nombre, edad):
        self.nombre = nombre
        self.edad = edad

    @abstractmethod
    def calcular_salario(self):
        pass

class Asalariado(Empleado):
    def __init__(self, nombre, edad, salario):
        super().__init__(nombre, edad)
        self.salario = salario

    def calcular_salario(self):
        return self.salario

class PorHora(Empleado):
    def __init__(self, nombre, edad, horas, dinero_por_hora):
        super().__init__(nombre, edad)
        self.horas = horas
        self.dinero_por_hora = dinero_por_hora

    def calcular_salario(self):
        return self.horas * self.dinero_por_hora

class Comisionista(Empleado):
    def __init__(self, nombre, edad, comision, trabajo):
        super().__init__(nombre, edad)
        self.comision = comision
        self.trabajo = trabajo

    def calcular_salario(self):
        return self.trabajo * self.comision

def AgregarAsalariado(empleados, indice):
    nombre = input("Ingrese el Nombre: ")
    edad = int(input("Ingrese la Edad: "))
    salario = int(input("Ingrese el Salario: "))
    empleados.append(Asalariado(nombre, edad, salario))

def AgregarPorHora(empleados, indice):
    nombre = input("Ingrese el Nombre: ")
    edad = int(input("Ingrese la Edad: "))
    horas = int(input("Ingrese las horas trabajadas: "))
    dineroxhora = float(input("Ingrese la Tarifa de dinero por hora: "))
    empleados.append(PorHora(nombre, edad, horas, dineroxhora))

def AgregarComisionista(empleados, indice):
    nombre = input("Ingrese el Nombre: ")
    edad = int(input("Ingrese la Edad: "))
    comision = float(input("Ingrese la Comision: "))
    trabajo = int(input("Ingrese la Meta de trabajo: "))
    empleados.append(Comisionista(nombre, edad, comision, trabajo))

def hayEmpleadosAgregados(empleados):
    return len(empleados) > 0

def mostrarInformacion(empleados):
    if not hayEmpleadosAgregados(empleados):
        print("Todavía no hay empleados registrados")
    else:
        for idx, empleado in enumerate(empleados, start=1):
            print(f"Información del Empleado {idx}:")
            print(f"Nombre: {empleado.nombre}")
            print(f"Edad: {empleado.edad}")
            print(f"Salario: {empleado.calcular_salario()}")

def BorrarEmpleados(empleados):
    if hayEmpleadosAgregados(empleados):
        empleados.clear()
        print("Empleados Borrados Exitosamente")
    else:
        print("No hay todavía empleados agregados")

def GestionarEmpleados():
    empleados = []
    while True:
        print("\nMenu de opciones:")
        print("1) Agregar Asalariado")
        print("2) Agregar Por Hora")
        print("3) Agregar Comisionista")
        print("4) Mostrar información de todos los empleados")
        print("5) Borrar Empleados")
        print("6) Salir")
        opcion = int(input("Seleccione una opción: "))

        if opcion == 1:
            AgregarAsalariado(empleados, len(empleados))
        elif opcion == 2:
            AgregarPorHora(empleados, len(empleados))
        elif opcion == 3:
            AgregarComisionista(empleados, len(empleados))
        elif opcion == 4:
            mostrarInformacion(empleados)
        elif opcion == 5:
            BorrarEmpleados(empleados)
        elif opcion == 6:
            break
        else:
            print("Opción no válida")

GestionarEmpleados()

File: Ejercicios/Ejercicio3/3.cpp
#include <iostream>
#include <string>
using namespace std;

class Empleado
{
public:
    string nombre;
    int edad;

    Empleado(string _nombre, int _edad) : nombre(_nombre), edad(_edad) {}
    virtual double CalcularSalario() const = 0;
    virtual ~Empleado() = default;
};

class Asalariado : public Empleado
{
private:
    int salario;

public:
    Asalariado(string _nombre, int _edad, int _salario) : Empleado(_nombre, _edad), salario(_salario) {}

    double CalcularSalario() const override // Polimorfismo
    {
        return salario;
    }
};

class PorHora : public Empleado
{
private:
    int horas;
    double dineroxhora;

public:
    PorHora(string _nombre, int _edad, int _horas, double _dineroxhora) : Empleado(_nombre, _edad), horas(_horas), dineroxhora(_dineroxhora) {}

    double CalcularSalario() const override // Polimorfismo
    {
        return horas * dineroxhora;
    }
};

class Comisionista : public Empleado
{
private:
    int trabajo;
    float comision;

public:
    Comisionista(string _nombre, int _edad, float _comision, int _trabajo) : Empleado(_nombre, _edad), comision(_comision), trabajo(_trabajo) {}

    double CalcularSalario() const override // Polimorfismo
    {
        return trabajo * comision;
    }
};

void AgregarAsalariado(Empleado* empleados[], int& indice)
{
    string nombre;
    int edad, salario;
    cout << "Ingrese el Nombre: " << endl;
    cin >> nombre;
    cout << "Ingrese la Edad: " << endl;
    cin >> edad;
    cout << "Ingrese el Salario: " << endl;
    cin >> salario;
    empleados[indice++] = new Asalariado(nombre, edad, salario);
}

void AgregarPorHora(Empleado* empleados[], int& indice)
{
    string nombre;
    int edad, horas;
    double dineroxhora;
    cout << "Ingrese el Nombre: " << endl;
    cin >> nombre;
    cout << "Ingrese la Edad: " << endl;
    cin >> edad;
    cout << "Ingrese las horas trabajadas: " << endl;
    cin >> horas;
    cout << "Ingrese la Tarifa de dinero por hora: " << endl;
    cin >> dineroxhora;
    empleados[indice++] = new PorHora(nombre, edad, horas, dineroxhora);
}

void AgregarComisionista(Empleado* empleados[], int& indice)
{
    string nombre;
    int edad, trabajo;
    float comision;
    cout << "Ingrese el Nombre: " << endl;
    cin >> nombre;
    cout << "Ingrese la Edad: " << endl;
    cin >> edad;
    cout << "Ingrese la Meta de trabajo: " << endl;
    cin >> trabajo;
    cout << "Ingrese la Comision: " << endl;
    cin >> comision;
    empleados[indice++] = new Comisionista(nombre, edad, comision, trabajo);
}

bool hayEmpleadosAgregados(Empleado* empleados[], int cantidad) // Verificamos si existen empleados en el arreglo
{
    for (int i = 0; i < cantidad; ++i)
    {
        if (empleados[i] != nullptr)
        {
            return true;
        }
    }
    return false;
}

void mostrarInformacion(Empleado* empleados[], int cantidad)
{
    if (!hayEmpleadosAgregados(empleados, cantidad))
    {
        cout << "Todavia no hay empleados registrados" << endl;
    }
    else
    {
        for (int i = 0; i < cantidad; ++i)
        {
            if (empleados[i] != nullptr)
            {
                cout << "Informacion del Empleado " << i + 1 << ":" << endl;
                cout << "Nombre: " << empleados[i]->nombre << endl;
                cout << "Edad: " << empleados[i]->edad << endl;
                cout << "Salario: " << empleados[i]->CalcularSalario() << endl;
            }
        }
    }
}

void borrarEmpleados(Empleado* empleados[], int cantidad)
{
    if (hayEmpleadosAgregados(empleados, cantidad))
    {
        for (int i = 0; i < cantidad; ++i)
        {
            delete empleados[i];
            empleados[i] = nullptr;
        }
        cout << "Empleados Borrados Exitosamente" << endl;
    }
    else
    {
        cout << "No hay todavia empleados agregados" << endl;
    }
}

void gestionarEmpleados()
{
    Empleado* empleados[100] = { nullptr }; // Inicializamos todos los punteros a nullptr
    int cantidadEmpleados = 0;
    int opcion;

    do
    {
        cout << "\nMenu de opciones:\n";
        cout << "1) Agregar Asalariado\n";
        cout << "2) Agregar PorHora\n";
        cout << "3) Agregar Comisionista\n";
        cout << "4) Mostrar informacion de todos los empleados\n";
        cout << "5) Borrar Empleados\n";
        cout << "6) Salir\n";
        cout << "Seleccione una opcion: ";
        cin >> opcion;

        switch (opcion)
        {
        case 1:
            AgregarAsalariado(empleados, cantidadEmpleados);
            break;
        case 2:
            AgregarPorHora(empleados, cantidadEmpleados);
            break;
        case 3:
            AgregarComisionista(empleados, cantidadEmpleados);
            break;
        case 4:
            mostrarInformacion(empleados, cantidadEmpleados);
            break;
        case 5:
            borrarEmpleados(empleados, cantidadEmpleados);
            cantidadEmpleados = 0; // Reiniciamos el contador
            break;
        case 6:
            borrarEmpleados(empleados, cantidadEmpleados); // Asegurarse de liberar memoria antes de salir
            break;
        default:
            cout << "Opcion no valida\n";
        }

    } while (opcion != 6);
}

int main()
{
    gestionarEmpleados();

    return 0;
}
